[narwhalsilent] iP

narwhalsilent opened the pull request:
May I interest thee in [this music](https://youtu.be/UNsrdjzLrzM?si=l3tvMoaUGCn5KhgS) while reading the following.

Greetings, Your Excellency :wine_glass:! Thy humble **Knight** doth stand before thee. I shall unburden thy mind of the weight of countless thoughts through meticulous task management.

> 'Tis befitting of a noble gentleman that tasks be managed by other minds.

Thy servant is:

- `command line` based
- Simple to use, as swift as a falcon in flight
- *Speedy* in its service to thee

To summon my aid, thou needest only:

1. Download my latest version.
2. Double-click it, like the forging of a mighty sword.
3. Add thy tasks, and let me, as a loyal squire, manage them for thee.

For thy fame and gallantry, I shall extend this service to thee without cost. 

Features include:
- [X] The management of tasks
- [ ] The oversight of impending deadlines (coming soon)
- [ ] Gentle reminders to aid thy noble pursuits (coming soon)

I have been learning the sorcery of coding, Your Excellency. May I demonstrate to thee:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

narwhalsilent committed:
Rename, Greet, Exit

narwhalsilent committed:
Echo

narwhalsilent committed:
Add, List

narwhalsilent committed:
Mark as Done

narwhalsilent committed:
ToDos, Events, Deadlines

narwhalsilent committed:
Automated Text UI Testing

narwhalsilent committed:
Handle Errors

narwhalsilent committed:
Delete

narwhalsilent committed:
Add save command

narwhalsilent committed:
Fix save command

narwhalsilent committed:
Add support for date as LocalDate

narwhalsilent committed:
Organise classes into package

narwhalsilent committed:
Merge branch 'add-gradle-support'

narwhalsilent committed:
Add JUnit Tests

narwhalsilent committed:
Add JavaDoc

narwhalsilent committed:
Add find method

narwhalsilent committed:
Comply with coding standard

narwhalsilent committed:
Merge branch 'branch-A-CodingStandard'

narwhalsilent committed:
Merge branch 'branch-Level-9'

narwhalsilent committed:
Update README.md

taufiq commented on src/main/java/knight/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+            return Command.MARK;
+        } else if (s.matches("unmark [1-9]\\d*")) {
+            return Command.UNMARK;
+        } else if (s.matches("delete [1-9]\\d*")) {
```
Cool use of regex to parse commands! Didn't think of that for mine

taufiq commented on src/main/java/knight/Knight.java:
```diff
@@ -0,0 +1,41 @@
+                + Ui.logo
+                + "doth stand before thee. How may I serve thee on this day?");
+
+        Scanner scan = new Scanner(System.in);
+        String message;
```
Remember to close your scanner

taufiq replied:
Remember to close your scanner

taufiq commented on src/main/java/knight/Storage.java:
```diff
@@ -0,0 +1,57 @@
+                }
+                tasks.executeCommandSilently(command, line);
+            }
+        } catch (NonstandardCommandException | DateTimeParseException e) {
```
Oh wow didn't know you could do that with the `|`. Cool!

taufiq replied:
Oh wow didn't know you could do that with the |. Cool!

taufiq commented on src/main/java/knight/ToDo.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Represents a task that needs to be done.
+ */
+public class ToDo extends Task {
```
I believe ToDo should be named as `Todo`? (According to the cs2103 convention that used the example of `Id` and not `ID`).

taufiq replied:
I believe ToDo should be named as Todo? (According to the cs2103 convention that used the example of Id and not ID).

taufiq commented on src/main/java/knight/TaskList.java:
```diff
@@ -0,0 +1,196 @@
+     * @return The commands representation of the task list.
+     */
+    String getCommands() {
+        String output = "";
```
Use a StringBuilder!

taufiq reviewed (approved):
I noticed that there's a lot of `try|catch` clauses scattered across the code. I think that works but you might want to consider just throwing the error and than catching it on a higher level (i.e the main loop). Would help to make your code look cleaner.

Sick use of regex and I learnt about the fact you could use `|` to union error types in the catch clause. Overall looks great! Thou be smashin' this iP 🔥

taufiq commented on src/main/java/knight/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+            return Command.SAVE;
+        } else if (s.equals("list")) {
+            return Command.LIST;
+        } else if (s.matches("mark [1-9]\\d*")) {
+            return Command.MARK;
+        } else if (s.matches("unmark [1-9]\\d*")) {
+            return Command.UNMARK;
+        } else if (s.matches("delete [1-9]\\d*")) {
+            return Command.DELETE;
+        } else if (s.matches("todo \\S.*")) {
+            return Command.TODO;
+        } else if (s.matches("deadline \\S.* /by \\S.*")) {
+            return Command.DEADLINE;
+        } else if (s.matches("event \\S.* /from \\S.* /to \\S.*")) {
+            return Command.EVENT;
+        } else if (s.matches("find \\S.*")) {
+            return Command.FIND;
```
Cool use of regex to parse commands! Didn't think of that for mine

justincred replied:
yes agree, well done on using regex for each command!

taufiq commented on src/main/java/knight/TaskList.java:
```diff
@@ -0,0 +1,196 @@
+     *
+     * @return The commands representation of the task list.
+     */
+    String getCommands() {
+        String output = "";
+        for (int i = 0; i < tasks.size(); i++) {
+            Task task = tasks.get(i);
+            if (i != tasks.size() - 1) {
+                output += task.getCommand() + "\n";
+                if (task.isDone) {
+                    output += "mark " + (i + 1) + "\n";
+                }
+            } else {
+                output += task.getCommand();
+                if (task.isDone) {
+                    output += "\nmark " + (i + 1);
+                }
+            }
+        }
+        return output;
```
Use a StringBuilder!

taufiq reviewed (commented):
I noticed that there's a lot of try|catch clauses scattered across the code. I think that works but you might want to consider just throwing the error and than catching it on a higher level (i.e the main loop). Would help to make your code look cleaner.

Sick use of regex and I learnt about the fact you could use | to union error types in the catch clause. Overall looks great! Thou be smashin' this iP 🔥

narwhalsilent committed:
Conform to coding standard

justincred commented on src/main/java/knight/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+            throw new NonstandardCommandException(
+                    "Thou shouldst forge a deadline task such as so:\n" +
+                            "deadline [description] /by [time]");
+        } else if (s.startsWith("event")) {
```
very thorough and interesting parsing of inputs!

justincred reviewed (commented):
Very unique and cool chatbot and not much to nitpick in terms of code quality!

narwhalsilent committed:
Implement GUI

Jolonauh commented on src/main/java/knight/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+     * @return The command representation of the task.
+     */
+    @Override
+    String getCommand() {
```
Remember to add access modifiers

Jolonauh commented on src/main/java/knight/Event.java:
```diff
@@ -0,0 +1,32 @@
+     * @return The command representation of the task.
+     */
+    @Override
+    String getCommand() {
```
Remember to add access modifiers

Jolonauh commented on src/main/java/knight/Task.java:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Mark the task as done.
+     */
+    void mark() {
```
Remember to add access modifiers

Jolonauh commented on src/main/java/knight/TaskList.java:
```diff
@@ -0,0 +1,299 @@
+     * @param commandType The type of command to execute.
+     * @param message     The full command message.
+     */
+    void executeCommand(Command commandType, String message) {
```
Remember to add access modifiers

Jolonauh commented on src/main/java/knight/Ui.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Controls the user interface of the program.
+ */
+public class Ui {
```
I like how you didn't capitalize abbreviations in class names

Jolonauh reviewed (commented):
Overall, the code mostly follows the code quality guidelines. You can consider creating more folders to organize your files (e.g. a folder for your task files).

narwhalsilent committed:
Improve code quality

narwhalsilent committed:
Improve code quality 2

narwhalsilent committed:
Improve code quality 3

narwhalsilent committed:
Add assertions

Add assertions to ensure program correctness

- Check that `CommandParser` works correctly.
- Check that `Event` and `Deadline` tasks cannot be successfully created with invalid time format.

narwhalsilent committed:
Merge pull request #2 from narwhalsilent/branch-A-Assertions

Add assertions

narwhalsilent committed:
Merge branch 'master' into branch-A-CodeQuality

narwhalsilent committed:
Merge pull request #1 from narwhalsilent/branch-A-CodeQuality

Improve code quality

narwhalsilent committed:
Implement update command

narwhalsilent committed:
Add Ui.png

narwhalsilent committed:
Update README.MD

narwhalsilent committed:
Update README.md again

narwhalsilent committed:
Update README.md yet again

narwhalsilent committed:
Update README.md yet yet again

narwhalsilent committed:
Refactor TaskList

Important updates:
- Refactor long methods in `TaskList`
- Reuse repeated code blocks from similar methods.

Minor updates:
- Improved Javadoc

narwhalsilent committed:
Update of README.md and implement `bye`

narwhalsilent committed:
Sync README.md
